### PR TITLE
Cutlass

### PR DESCRIPTION
**Pull request by JasonDing0401**
complex kernel works, but still in column major. All changes are in `kaasComplex/`.

### CODE REVIEW

File: examples/cutlass/kaasComplex/Makefile [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+# 75FOUND:=$(shell lspci | grep "T4")
+
+# ifneq ($(75FOUND), )
+# 	COMPUTE_CAP:=75
+# else
+# 	COMPUTE_CAP:=35
+# endif
+
+COMPUTE_CAP:=35
```
Comment (NathanTP) [inline examples/cutlass/kaasComplex/Makefile:11]: gotta use the commented version above that detects the GPU type. We need to be able to run this on a V100 (ec2) as well.

File: examples/cutlass/build.sh [hunk 2]
```diff
@@ -27,3 +27,41 @@ CUTLASS_PATH=$(realpath $CUTLASS_PATH)
 #     -lcudadevrt -lcudart_static -lrt -lpthread -ldl \
 #     -L"/usr/local/cuda/targets/x86_64-linux/lib/stubs" \
 #     -L"/usr/local/cuda/targets/x86_64-linux/lib"
+
+/usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler \
+-I/scratch/jasonding/fakefaas/examples/cutlass/cutlass/include \
```
Comment (NathanTP) [inline examples/cutlass/build.sh]: your directory is hard-coded in here. Better to use the $CUTLASS_PATH variable that the first command uses.

File: examples/cutlass/kaasComplex/test.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+def testSgemm(M, N, K, alpha, beta):
+    lda = M
+    ldb = K
+    ldc = M
+
+    getArg, getDims = loadAdapter()
+    refKern, cutlassKern = loadKerns()
+
+    rng = np.random.default_rng(5)
+    a = np.asfortranarray(rng.random((M, K), dtype=np.float32) * (1+1j))
```
Comment (NathanTP) [inline examples/cutlass/kaasComplex/test.py]: not all that important I guess, but we may as well make completely random complex numbers. What's here will have a random real part with a non-random imaginary part. Try this instead:
np.asfortranarray(rng.random((M, K), dtype=np.float32) * (rng.random((M,K), dtype=np.float32)*1j))

**Review by NathanTP — COMMENTED** [review]
Eventually we're gonna need to clean this directory up a bit, it's hard to tell what's going on from the directory names and I'm pretty sure build.sh is out of date. For now, it's enough to have it serve as a starting point for the kaas integration.

Another issue: I tried this out and the numpy and cutlass results seem to differ in the latter part of the result matrix. Numpy has all zeros for the real part which seems sort of suspicious. Do we know what's going on there?

Also the reference kernel seems completely wrong, but I honestly don't think that's worth spending time on, we only care if numpy and cutlass agree.

File: examples/cutlass/build.sh (hunk 2 above)
Comment (JasonDing0401) [reply examples/cutlass/build.sh]: I have makefile inside both gemm and complex gemm, so I think I should delete this build.sh?

**Comment by JasonDing0401** [comment]
```
Grid is:  (1, 1, 1)
Block is:  (256, 1, 1)
Smem Size is:  33024
Cutlass Kern Result:
[[-0.88970816+61.60884j   0.8753367 +66.304985j -1.2602694 +56.77777j
  ...  3.3594985 +61.968437j  0.06158584+63.717854j
   1.3220109 +63.38286j ]
 [ 4.1345954 +59.66194j   1.4608327 +64.2702j    1.8210521 +58.569164j
  ...  4.0277643 +62.317135j  4.395748  +64.00689j
   6.631257  +60.352734j]
 [ 1.2082844 +60.352642j -0.38606754+61.026405j -0.3192921 +56.527287j
  ...  0.80580825+60.548584j  1.2763513 +63.674927j
   5.2991557 +60.734478j]
 ...
 [ 2.188159  +65.68774j  -0.16279075+68.53394j  -2.1837013 +61.675095j
  ...  2.1400197 +65.039444j  0.10705746+70.17587j
   3.6032631 +64.63495j ]
 [-0.49709833+66.41434j   0.7009518 +70.855934j -2.046655  +62.691814j
  ...  0.44279948+67.23151j  -0.2491381 +69.72685j
   1.8205959 +67.40263j ]
 [ 2.2860558 +59.27939j   1.0868267 +63.129887j -0.2808098 +53.25511j
  ...  0.16780984+56.574146j -0.9851943 +60.313793j
   2.753599  +59.19283j ]]
Reference Kern Result:
[[-0.88970995+61.608826j  4.134596  +59.66195j   1.2082856 +60.352665j
  ...  2.1881585 +65.68775j  -0.497098  +66.41432j
   2.2860556 +59.279408j]
 [ 0.8753368 +66.30499j   1.460833  +64.27021j  -0.3860676 +61.026382j
  ... -0.16279069+68.53394j   0.7009506 +70.855934j
   1.0868253 +63.129894j]
 [-1.2602693 +56.777786j  1.8210534 +58.569183j -0.31929228+56.527287j
  ... -2.183702  +61.67508j  -2.0466545 +62.691803j
  -0.2808113 +53.255127j]
 ...
 [ 3.3594978 +61.968437j  4.027766  +62.31715j   0.8058078 +60.548553j
  ...  2.1400204 +65.03945j   0.44280005+67.23147j
   0.16781092+56.57416j ]
 [ 0.06158553+63.717854j  4.3957505 +64.00691j   1.2763492 +63.674915j
  ...  0.10705644+70.175865j -0.24913791+69.72687j
  -0.98519486+60.31381j ]
 [ 1.3220106 +63.38285j   6.6312547 +60.352757j  5.299154  +60.734474j
  ...  3.6032617 +64.63492j   1.8205955 +67.40266j
   2.7536001 +59.192833j]]
NP Result: 
[[-0.88970816+61.60884j   0.8753367 +66.304985j -1.2602694 +56.77777j
  ...  3.3594985 +61.968437j  0.06158584+63.717854j
   1.3220109 +63.38286j ]
 [ 4.1345954 +59.66194j   1.4608327 +64.2702j    1.8210521 +58.569164j
  ...  4.0277643 +62.317135j  4.395748  +64.00689j
   6.631257  +60.352734j]
 [ 1.2082844 +60.352642j -0.38606754+61.026405j -0.3192921 +56.527287j
  ...  0.80580825+60.548584j  1.2763513 +63.674927j
   5.2991557 +60.734478j]
 ...
 [ 2.188159  +65.68774j  -0.16279075+68.53394j  -2.1837013 +61.675095j
  ...  2.1400337 +65.03947j   0.10706329+70.17587j
   3.6032658 +64.63494j ]
 [-0.49709833+66.41434j   0.7009518 +70.855934j -2.046655  +62.691814j
  ...  0.44280243+67.23149j  -0.24913025+69.72687j
   1.8205948 +67.40264j ]
 [ 2.2860558 +59.27939j   1.0868267 +63.129887j -0.2808098 +53.25511j
  ...  0.16780663+56.574165j -0.98518753+60.313797j
   2.7535915 +59.19282j ]]
Difference between numpy and reference result:
[[ 1.7881393e-06+1.5258789e-05j -3.2592592e+00+6.6430359e+00j
  -2.4685550e+00-3.5748940e+00j ...  1.1713400e+00-3.7193146e+00j
   5.5868381e-01-2.6964684e+00j -9.6404469e-01+4.1034508e+00j]
 [ 3.2592585e+00-6.6430511e+00j -2.3841858e-07-7.6293945e-06j
   2.2071197e+00-2.4572182e+00j ...  4.1905551e+00-6.2168083e+00j
   3.6947975e+00-6.8490448e+00j  5.5444317e+00-2.7771606e+00j]
 [ 2.4685535e+00+3.5748558e+00j -2.2071209e+00+2.4572220e+00j
   1.7881393e-07+0.0000000e+00j ...  2.9895103e+00-1.1264954e+00j
   3.3230057e+00+9.8312378e-01j  5.5799670e+00+7.4793510e+00j]
 ...
 [-1.1713388e+00+3.7192993e+00j -4.1905570e+00+6.2167931e+00j
  -2.9895091e+00+1.1265411e+00j ...  1.3351440e-05+1.5258789e-05j
  -3.3573675e-01+2.9444046e+00j  3.4354548e+00+8.0607796e+00j]
 [-5.5868387e-01+2.6964836e+00j -3.6947987e+00+6.8490219e+00j
  -3.3230042e+00-9.8310089e-01j ...  3.3574599e-01-2.9443741e+00j
   7.6591969e-06+0.0000000e+00j  2.8057897e+00+7.0888329e+00j]
 [ 9.6404517e-01-4.1034622e+00j -5.5444279e+00+2.7771301e+00j
  -5.5799637e+00-7.4793625e+00j ... -3.4354551e+00-8.0607529e+00j
  -2.8057830e+00-7.0888596e+00j -8.5830688e-06-1.1444092e-05j]]
Difference between cutlass and numpy result:
[[ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ...  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j]
 [ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ...  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j]
 [ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ...  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j]
 ...
 [ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ...  1.4066696e-05+2.2888184e-05j
   5.8338046e-06+0.0000000e+00j  2.6226044e-06-7.6293945e-06j]
 [ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ...  2.9504299e-06-1.5258789e-05j
   7.8529119e-06+1.5258789e-05j -1.0728836e-06+7.6293945e-06j]
 [ 0.0000000e+00+0.0000000e+00j  0.0000000e+00+0.0000000e+00j
   0.0000000e+00+0.0000000e+00j ... -3.2186508e-06+1.9073486e-05j
   6.7949295e-06+3.8146973e-06j -7.3909760e-06-7.6293945e-06j]]
```
This is the result I get. I don't know `Numpy has all zeros for the real part`? I think the difference of `np - cutlass` is good? The locations where the difference is not zero has really small value.

**Comment by JasonDing0401** [comment]
Another quick question: cuda nvcc works for version 10.2 but not version 11+, so I kind of hard code it in the makefile. How could I change that to adapt to other computers?

**Comment by JasonDing0401** [comment]
The reference kernel bug is fixed. I also polish the code a little bit.

File: examples/cutlass/build.sh (hunk 2 above)
Comment (NathanTP) [reply examples/cutlass/build.sh]: yes, deleting it is fine

**Comment by JasonDing0401** [comment]
NVCC should now work for both version 10 and 11.

**Review by NathanTP — APPROVED** [review]
seems to work on my side

**Pull request merged**